x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java

```diff
@@ -384,6 +416,16 @@ void refresh(PersistentTasksCustomMetadata persistentTasks, ActionListener<Void>
         }
     }
 
+    private void refreshTrainedModelTasks(PersistentTasksCustomMetadata persistentTasks) {
+        memoryRequirementByTrainedModelTask.clear();
```

Comment:
is there a reason that it's cleared and than recalculated? Seems potentially buggy to me, even if currently the access pattern might be safe. The type is a _concurrent_ hashmap, so indicates concurrent access. In an unlucky moment it returns a wrong result.

Why not calculating it fresh and reassigning it?
(^ this is tricky, it invalidates `final` and might be unsafe, too. You could loop through the tasks and just update the values and at the same time remember the seen tasks in a Set. Afterwards you loop again and delete entries for unseen tasks.)

Reply:
This is a good point I think. The reason I clear it is to avoid having to manually remove entries from there when the tasks stop. But I think you're right that a read might come in halfway through a refresh and hit a missing value for a running job. I'll think of a way to solve it.